### PR TITLE
Bug 1806899 - Detect link escapes in safe_extract

### PR DESCRIPTION
**Pull request by gbrownmozilla**

### CODE REVIEW

**Comment by gbrownmozilla** [comment]
@jcristau I recall you said this approach wouldn't always work, but I can't think of why, and it detects these test cases.

**Review by jcristau — CHANGES_REQUESTED** [review]
I believe this rejects harmless symlinks and potentially doesn't catch ones that do escape.

File: client/tooltool.py [hunk 1]
```diff
@@ -958,14 +958,18 @@ def clean_path(dirname):
...
+        real_directory = os.path.realpath(directory)
+        real_target = os.path.realpath(target)
+        prefix = os.path.commonprefix([real_directory, real_target])
+        return prefix == real_directory
 
     member_path = os.path.join(path, member.name)
     if not _is_within_directory(path, member_path):
         raise Exception("Attempted path traversal in tar file: " + member.name)
+    if member.issym():
+        link_path = os.path.join(path, member.linkname)
```
[the remark this answers is not in the document]
Comment (gbrownmozilla) [reply client/tooltool.py:970]: Yes, thanks for catching that.

**Comment by gbrownmozilla** [comment]
Sorry, messed up the merge...let's just start over...

**Pull request closed without merging**